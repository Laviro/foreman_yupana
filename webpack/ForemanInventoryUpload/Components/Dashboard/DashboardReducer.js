import Immutable from 'seamless-immutable';

import {
  INVENTORY_POLLING_START,
  INVENTORY_POLLING,
  INVENTORY_TAB_CHANGED,
  INVENTORY_POLLING_ERROR,
} from './DashboardConstants';

const initialState = Immutable({});

export default (state = initialState, action) => {
  const {
    payload: { accountID, pollingProcessID, logs, error, activeTab } = {},
  } = action;
  switch (action.type) {
    case INVENTORY_POLLING_START:
      return state.setIn([accountID], {
        ...state[accountID],
        pollingProcessID,
        activeTab: 'generating',
      });
    case INVENTORY_POLLING:
      return state.setIn([accountID], {
        ...state[accountID],
        [activeTab]: {
          logs,
          error: null,
        },
      });
    case INVENTORY_TAB_CHANGED:
      return state.setIn([accountID], {
        ...state[accountID],
        activeTab,
      });
    case INVENTORY_POLLING_ERROR:
      return state.setIn([accountID], {
        ...state[accountID],
        [activeTab]: {
          ...state[activeTab],
          error,
        },
      });
    default:
      return state;
  }
};
