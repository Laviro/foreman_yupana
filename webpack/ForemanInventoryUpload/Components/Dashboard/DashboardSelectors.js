import { selectForemanInventoryUpload } from '../../ForemanInventoryUploadSelectors';

export const selectDashboard = (state, accountID) =>
  selectForemanInventoryUpload(state).dashboard[accountID] || {};
export const selectUploading = (state, accountID) =>
  selectDashboard(state, accountID).uploading;
export const selectGenerating = (state, accountID) =>
  selectDashboard(state, accountID).generating;
export const selectPollingProcessID = (state, accountID) =>
  selectDashboard(state, accountID).pollingProcessID;
export const selectActiveTab = (state, accountID) =>
  selectDashboard(state, accountID).activeTab;
