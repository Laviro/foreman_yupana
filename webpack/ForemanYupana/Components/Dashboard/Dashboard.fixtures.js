import { seperator } from './DashboardHelper';

export const pollingProcessID = 1;

export const logs = ['some-logs...'];

export const completed = 25;

export const yupana = {
  generating: {
    logs,
    completed,
  },
  uploading: {
    logs,
    completed,
    files,
  },
};

export const files = ['some-file'];

export const activeTab = 'uploading';

const initialLog = ['No running process', seperator];

export const initialState = {
  generating: {
    logs: initialLog,
    completed,
  },
  uploading: {
    logs: initialLog,
    completed,
    files,
  },
  pollingProcessID: 0,
  activeTab,
};
