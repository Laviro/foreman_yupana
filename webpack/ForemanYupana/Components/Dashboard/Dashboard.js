import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';
import ReportGenerate from '../ReportGenerate';
import ReportUpload from '../ReportUpload';
import NavContainer from '../NavContainer';
import './dashboard.scss';

class Dashboard extends React.Component {
  componentDidMount() {
    const { startPolling, fetchLogs } = this.props;
    const pollingProcessID = setInterval(fetchLogs, 2000);
    startPolling(pollingProcessID);
  }

  componentWillUnmount() {
    const { stopPolling } = this.props;
    stopPolling();
  }

  render() {
    const { setActiveTab } = this.props;
    return (
      <NavContainer
        items={[
          {
            icon: 'database',
            name: 'Generating',
            component: ReportGenerate,
          },
          {
            icon: 'cloud-upload',
            name: 'Uploading',
            component: ReportUpload,
          },
        ]}
        onTabClick={setActiveTab}
      />
    );
  }
}

Dashboard.propTypes = {
  startPolling: PropTypes.func,
  fetchLogs: PropTypes.func,
  stopPolling: PropTypes.func,
  setActiveTab: PropTypes.func,
};

Dashboard.defaultProps = {
  startPolling: noop,
  fetchLogs: noop,
  stopPolling: noop,
  setActiveTab: noop,
};

export default Dashboard;
