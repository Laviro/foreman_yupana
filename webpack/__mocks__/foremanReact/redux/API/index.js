export const get = data => ({ type: 'get-some-type', ...data });
export const put = data => ({ type: 'put-some-type', ...data });
export const post = data => ({ type: 'post-some-type', ...data });
export const patch = data => ({ type: 'patch-some-type', ...data });

export const API = {
  get: jest.fn(),
  put: jest.fn(),
  post: jest.fn(),
  delete: jest.fn(),
  patch: jest.fn(),
};
