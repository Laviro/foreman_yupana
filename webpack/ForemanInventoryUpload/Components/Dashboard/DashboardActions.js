import API from 'foremanReact/API';
import { selectActiveTab } from './DashboardSelectors';
import { inventoryUrl } from '../../ForemanInventoryHelpers';
import {
  INVENTORY_POLLING_START,
  INVENTORY_POLLING_STOP,
  INVENTORY_POLLING,
  INVENTORY_TAB_CHANGED,
  INVENTORY_POLLING_ERROR,
  INVENTORY_REPORTS_DOWNLOAD,
  INVENTORY_TOGGLE_TERMINAL_FULL_SCREEN,
} from './DashboardConstants';

export const startPolling = (accountID, pollingProcessID) => ({
  type: INVENTORY_POLLING_START,
  payload: {
    accountID,
    pollingProcessID,
  },
});

export const stopPolling = (accountID, pollingProcessID) => dispatch => {
  clearInterval(pollingProcessID);
  dispatch({
    type: INVENTORY_POLLING_STOP,
    payload: {
      accountID,
    },
  });
};

export const fetchLogs = accountID => async (dispatch, getState) => {
  const activeTab = selectActiveTab(getState(), accountID);
  try {
    const processController = activeTab === 'uploading' ? 'uploads' : 'reports';
    const {
      data: { output, scheduled },
    } = await API.get(inventoryUrl(`${accountID}/${processController}/last`));
    const outputArray = output.split('\n');
    dispatch({
      type: INVENTORY_POLLING,
      payload: {
        accountID,
        activeTab,
        logs: outputArray,
        scheduled,
      },
    });
  } catch (error) {
    dispatch({
      type: INVENTORY_POLLING_ERROR,
      payload: {
        accountID,
        activeTab,
        error: error.message,
      },
    });
  }
};

export const setActiveTab = (accountID, tabName) => ({
  type: INVENTORY_TAB_CHANGED,
  payload: {
    accountID,
    activeTab: tabName,
  },
});

export const downloadReports = accountID => {
  window.location.href = inventoryUrl(`${accountID}/uploads/file`);
  return {
    type: INVENTORY_REPORTS_DOWNLOAD,
    payload: {
      accountID,
    },
  };
};

export const toggleFullScreen = accountID => (dispatch, getState) => {
  const activeTab = selectActiveTab(getState(), accountID);
  dispatch({
    type: INVENTORY_TOGGLE_TERMINAL_FULL_SCREEN,
    payload: {
      activeTab,
      accountID,
    },
  });
};
