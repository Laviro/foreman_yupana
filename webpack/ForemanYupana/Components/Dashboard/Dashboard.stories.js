import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import { Provider } from 'react-redux';
import configureStore from '../../../stories/configureStore';
import Dashboard from './index';
import { generating, uploading } from './Dashboard.fixtures';

const store = configureStore();

storiesOf('Dashboard', module)
  .addDecorator(withKnobs)
  .add('Dashboard', () => (
    <div style={{ margin: '20px' }}>
      <Provider store={store}>
        <Dashboard generating={generating} uploading={uploading} />
      </Provider>
    </div>
  ));
