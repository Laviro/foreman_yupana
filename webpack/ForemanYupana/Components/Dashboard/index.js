import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './DashboardActions';
import reducer from './DashboardReducer';

import Dashboard from './Dashboard';

// map state to props
const mapStateToProps = state => ({});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { dashboard: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Dashboard);
