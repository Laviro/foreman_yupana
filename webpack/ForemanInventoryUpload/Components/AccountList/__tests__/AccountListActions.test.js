import { testActionSnapshotWithFixtures } from 'react-redux-test-utils';
import API from 'foremanReact/API';
import {
  fetchAccountsStatus,
  startAccountStatusPolling,
  stopAccountStatusPolling,
  restartProcess,
} from '../AccountListActions';
import {
  pollingProcessID,
  fetchAccountsStatusResponse,
} from '../AccountList.fixtures';
import { accountID, activeTab } from '../../Dashboard/Dashboard.fixtures';

jest.mock('foremanReact/API');
API.get.mockImplementation(async () => fetchAccountsStatusResponse);

const fixtures = {
  'should fetchAccountsStatus': () => fetchAccountsStatus(),
  'should startAccountStatusPolling': () =>
    startAccountStatusPolling(pollingProcessID),
  'should stopAccountStatusPolling': () =>
    stopAccountStatusPolling(pollingProcessID),
  'should restartProcess': () => restartProcess(accountID, activeTab),
};

describe('AccountList actions', () => testActionSnapshotWithFixtures(fixtures));
