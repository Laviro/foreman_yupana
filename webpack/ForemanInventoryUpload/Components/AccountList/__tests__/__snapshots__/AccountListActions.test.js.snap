// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountList actions should fetchAccountsStatus 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "accounts": Object {
          "Account1": Object {
            "generate_report_status": "running",
            "label": "test_org1",
            "upload_report_status": "running",
          },
          "Account2": Object {
            "generate_report_status": "failure",
            "label": "test_org2",
            "upload_report_status": "unknown",
          },
          "Account3": Object {
            "generate_report_status": "running",
            "label": "test_org3",
            "upload_report_status": "success",
          },
        },
        "autoUploadEnabled": true,
        "hostObfuscationEnabled": true,
      },
      "type": "INVENTORY_ACCOUNT_STATUS_POLLING",
    },
  ],
]
`;

exports[`AccountList actions should restartProcess 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "accountID": "some-account-ID",
        "processStatusName": "generate_report_status",
      },
      "type": "INVENTORY_PROCESS_RESTART",
    },
  ],
]
`;

exports[`AccountList actions should startAccountStatusPolling 1`] = `
Object {
  "payload": Object {
    "pollingProcessID": 0,
  },
  "type": "INVENTORY_ACCOUNT_STATUS_POLLING_START",
}
`;

exports[`AccountList actions should stopAccountStatusPolling 1`] = `
Array [
  Array [
    Object {
      "type": "INVENTORY_ACCOUNT_STATUS_POLLING_STOP",
    },
  ],
]
`;
